akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRunning.scala

```diff
@@ -0,0 +1,363 @@
+ * - PersistingEvents - where incoming commands are stashed until persistence completes
+ *
+ * This is implemented as such to avoid creating many EventsourcedRunning instances,
+ * which perform the Persistence extension lookup on creation and similar things (config lookup)
+ *
+ */
+@InternalApi
+class EventsourcedRunning[Command, Event, State](
+  val setup:                  EventsourcedSetup[Command, Event, State],
+  override val context:       ActorContext[Any],
+  override val timers:        TimerScheduler[Any],
+  override val internalStash: StashBuffer[Any],
+
+  private var sequenceNr: Long,
+  val writerIdentity:     WriterIdentity,
+
+  private var state: State
+) extends MutableBehavior[Any]
+  with EventsourcedBehavior[Command, Event, State]
+  with EventsourcedStashManagement { same ⇒
+  import setup._
+
+  import EventsourcedBehavior._
+  import akka.actor.typed.scaladsl.adapter._
+
+  protected val log = Logging(context.system.toUntyped, this)
+
+  private def commandContext: ActorContext[Command] = context.asInstanceOf[ActorContext[Command]]
+
+  // ----------
+
+  // Holds callbacks for persist calls (note that we do not implement persistAsync currently)
+  private def hasNoPendingInvocations: Boolean = pendingInvocations.isEmpty
+  private val pendingInvocations = new java.util.LinkedList[PendingHandlerInvocation]() // we only append / isEmpty / get(0) on it
+
+  // ----------
+
+  private def snapshotSequenceNr: Long = sequenceNr
+
+  private def updateLastSequenceNr(persistent: PersistentRepr): Unit =
+    if (persistent.sequenceNr > sequenceNr) sequenceNr = persistent.sequenceNr
+  private def nextSequenceNr(): Long = {
+    sequenceNr += 1L
+    sequenceNr
+  }
+  // ----------
+
+  private def onSnapshotterResponse(response: SnapshotProtocol.Response): Behavior[Any] = {
+    response match {
+      case SaveSnapshotSuccess(meta) ⇒
+        log.debug("Save snapshot successful: " + meta)
+        same
+      case SaveSnapshotFailure(meta, ex) ⇒
+        log.error(ex, "Save snapshot failed! " + meta)
+        same // FIXME https://github.com/akka/akka/issues/24637 should we provide callback for this? to allow Stop
+    }
+  }
+
+  // ----------
+
+  trait EventsourcedRunningPhase {
+    def name: String
+    def onCommand(c: Command): Behavior[Any]
+    def onJournalResponse(response: JournalProtocol.Response): Behavior[Any]
+  }
+
+  object HandlingCommands extends EventsourcedRunningPhase {
+    def name = "HandlingCommands"
+
+    final override def onCommand(command: Command): Behavior[Any] = {
+      val effect = commandHandler(commandContext, state, command)
+      applyEffects(command, effect.asInstanceOf[EffectImpl[E, S]]) // TODO can we avoid the cast?
+    }
+    final override def onJournalResponse(response: Response): Behavior[Any] = {
+      // should not happen, what would it reply?
+      throw new RuntimeException("Received message which should not happen in Running state!")
```

Comment:
I think such can be possible when the actor is restarted. That is the reason for the instanceId, and such messages targeted to wrong/old instance should be discarded, not cause exception